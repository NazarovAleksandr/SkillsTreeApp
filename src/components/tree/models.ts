export interface ITreeNode {
    id: string;
    children?: ITreeNode[];
    isRoot?: boolean;
}

export interface ITreeNodeProps {
    node: ITreeNode;
    onSelect?: (selectedNode: ITreeNode) => void;
    nodeContent: (node: ITreeNode) => React.ReactNode;
}

export interface ITreeProps<T> {
    nodes: (ITreeNode & T)[];
    onSelect?: (selectedNode: ITreeNode & T) => void;
    children: (node: ITreeNode & T) => React.ReactNode;
}