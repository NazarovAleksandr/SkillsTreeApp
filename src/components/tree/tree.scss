.node-container {
    display: flex;
    align-items: center;

    .node {
        cursor: pointer;
    }
}