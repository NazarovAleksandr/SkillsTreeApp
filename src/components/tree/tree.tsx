import * as React from 'react';
import * as Models from './models';
import './tree.scss';
import { observer } from 'mobx-react';

export const TreeNode = observer((props: Models.ITreeNodeProps) => {
    const { node, onSelect, nodeContent } = props;
    const onClick = () => {
        onSelect && onSelect(node);
    };
    const userContent = nodeContent && nodeContent(node);

    return (
        <div className="node-container">
            <div className="node" onClick={onClick}>{userContent}</div>
            {
                node.children && node.children.length
                    && (
                        <div className="children-list">
                            {node.children.map((childNode, idx) => <TreeNode key={idx} node={childNode} onSelect={onSelect} nodeContent={nodeContent} />)}
                        </div>
                    )
            }
        </div>
    );
});

export const Tree = observer(<T extends {}>(props: Models.ITreeProps<T>) => {
    const { onSelect, nodes, children } = props;
    return (
        <div className="tree">
            {
                nodes.map((node, idx) => <TreeNode key={idx} node={node} onSelect={onSelect} nodeContent={children} />)
            }
        </div>
    );
});
