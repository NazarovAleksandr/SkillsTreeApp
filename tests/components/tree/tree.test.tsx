import * as React from 'react';
import { shallow } from 'enzyme';
import { Tree, TreeNode } from '../../../src/components/tree/tree';

describe('Tree', () => {
    test('Renders roots', () => {
        const tree = [{
            id: '1',
            children: [{ id: '2' }],
        }, {
            id: '3',
            children: [{ id: '4' }],
        }];
        const component = shallow(<Tree nodes={tree}>{() => (123)}</Tree>);

        expect(component).toMatchSnapshot('roots');
    });
    test('Renders TreeNode', () => {
        const tree = {
            id: '1',
            children: [{ id: '2' }, { id: '3' }],
        };
        const component = shallow(<TreeNode node={tree} nodeContent={(node) => (node.id)} />);

        expect(component).toMatchSnapshot('tree node');
    });
    test('Calls callback', () => {
        const tree = {
            id: '1',
            children: [{ id: '2' }, { id: '3' }],
        };
        const fn = jest.fn();
        const component = shallow(<TreeNode node={tree} onSelect={fn} nodeContent={(node) => (node.id)} />);
        component.find('.node').simulate('click');

        expect(fn).toBeCalled();
    });
});
