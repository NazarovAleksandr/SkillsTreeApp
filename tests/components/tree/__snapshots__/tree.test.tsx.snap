// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree Renders TreeNode: tree node 1`] = `
<div
  className="node-container"
>
  <div
    className="node"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    1
  </div>
  <div
    className="children-list"
  >
    <Memo(wrappedComponent)
      key="2"
      node={
        Object {
          "id": "2",
        }
      }
      nodeContent={[Function]}
    />
    <Memo(wrappedComponent)
      key="3"
      node={
        Object {
          "id": "3",
        }
      }
      nodeContent={[Function]}
    />
  </div>
</div>
`;

exports[`Tree Renders roots: roots 1`] = `
<div
  className="tree"
>
  <Memo(wrappedComponent)
    key="1"
    node={
      Object {
        "children": Array [
          Object {
            "id": "2",
          },
        ],
        "id": "1",
      }
    }
    nodeContent={[Function]}
  />
  <Memo(wrappedComponent)
    key="3"
    node={
      Object {
        "children": Array [
          Object {
            "id": "4",
          },
        ],
        "id": "3",
      }
    }
    nodeContent={[Function]}
  />
</div>
`;
